README.md

```diff
@@ -389,123 +407,71 @@ This `AuthenticityTokenInput` will get the authenticity token from the `Authenti
-
-### DynamicLinks
-
-> **Warning**: Deprecated in favor of the `V2_MetaFunction`. This will be removed in the next major version. Check below for the new way to do this.
-
-If you need to create `<link />` tags based on the loader data instead of being static, you can use the `DynamicLinks` component together with the `DynamicLinksFunction` type.
-
-In the route you want to define dynamic links add `handle` export with a `dynamicLinks` method, this method should implement the `DynamicLinksFunction` type.
+If you need to parse the body as FormData yourself (e.g. to support file uploads) you can also call `CSRF#validate` with the FormData and Headers objects.
 
 ```ts
-// create the dynamicLinks function with the correct type
-// note: loader type is optional
-let dynamicLinks: DynamicLinksFunction<SerializeFrom<typeof loader>> = ({
-  id,
-  data,
-  params,
-  matches,
-  location,
-  parentsData,
-}) => {
-  if (!data.user) return [];
-  return [{ rel: "preload", href: data.user.avatar, as: "image" }];
-};
-
-// and export it through the handle, you could also create it inline here
-// if you don't care about the type
-export let handle = { dynamicLinks };
-```
-
-Then, in the root route, add the `DynamicLinks` component before the Remix's Links component, usually inside a Document component.
-
-```tsx
-import { Links, LiveReload, Meta, Scripts, ScrollRestoration } from "remix";
-import { DynamicLinks } from "remix-utils";
-
-type Props = { children: React.ReactNode; title?: string };
-
-export function Document({ children, title }: Props) {
-  return (
-    <html lang="en">
-      <head>
-        <meta charSet="utf-8" />
-        <meta name="viewport" content="width=device-width,initial-scale=1" />
-        {title ? <title>{title}</title> : null}
-        <Meta />
-        <DynamicLinks />
-        <Links />
-      </head>
-      <body>
-        {children}
-        <ScrollRestoration />
-        <Scripts />
-        <LiveReload />
-      </body>
-    </html>
-  );
+let formData = await parseMultiPartFormData(request);
+try {
+  await csrf.validate(formData, request.headers);
+} catch (error) {
+  // handle errors
 }
 ```
 
-Now, any link you defined in the `DynamicLinksFunction` will be added to the HTML as any static link in your `LinksFunction`s.
+> **Warning**: If you call `CSRF#validate` with the request instance but you read the body before it will throw an error.
 
-> **Note**
-> You can also put the `DynamicLinks` after the `Links` component, it's up to you what to prioritize, since static links are probably prefetched when you do `<Link prefetch>` you may want to put the `DynamicLinks` first to prioritize them.
+In case the CSRF validation fails, it will throw a `CSRFError` which can be used to correctly identify them against other possible errors that may throw.
```

Comment:
```suggestion
In case the CSRF validation fails, it will throw a `CSRFError` which can be used to correctly identify it against other possible errors that may get thrown.
```